state_machine/drone.py

```diff
@@ -110,6 +113,65 @@ async def connect_drone(self) -> None:
             else dronekit.connect(self.address, wait_ready=True, baud=self.baud)
         )
 
+    async def arm_and_takeoff(self, takeoff_alt: float) -> None:
+        """
+        Arm the drone and takeoff vertically to passed altitude
+
+        Parameters
+        ----------
+        takeoff_alt: float
+            Altitude to reach in meters
+        """
+        while not self.vehicle.is_armable:
+            # Vehicle is not ready to accept code
+            logging.info("Waiting for vehicle to initialize...")
+
+        self.vehicle.mode = dronekit.VehicleMode("GUIDED")
+        self.vehicle.armed = True
+
+        # Confirm vehicle is properly armed
+        while not self.vehicle.armed:
+            logging.info("Waiting for arming...")
+            await asyncio.sleep(1)
+
+        self.vehicle.simple_takeoff(takeoff_alt)
+
+        # Verify vehicle reaches target altitude
+        while (
+            self.vehicle.location.global_relative_frame.alt <= takeoff_alt * 0.95
+        ):  # 95% takeoff alt for margin of error
+            await asyncio.sleep(0.5)
+        logging.info(f"Reached target altitude ({takeoff_alt} m).")
+        return
+
+    async def return_to_launch(self) -> None:
+        """
+        Method to move vehicle above home location, then descend vertically
+        """
+        home_loc = dronekit.LocationGlobalRelative(
+            self.vehicle.home_location.lat, self.vehicle.home_location.lon, 23
+        )  # Min alt should be in constants file
+        self.vehicle.simple_goto(home_loc)
+        logging.info("Moving to home location...")
+        while (
+            calculate_distance(
+                self.vehicle.location.global_relative_frame.lat,
+                self.vehicle.location.global_relative_frame.lon,
+                self.vehicle.location.global_relative_frame.alt,
+                home_loc.lat,
+                home_loc.lon,
+                home_loc.alt,
+            )
+            > 1
+        ):  # Get within 1 meter above home location
+            await asyncio.sleep(0.5)
+        self.vehicle.mode = dronekit.VehicleMode("RTL")
+        logging.info("Returning to launch...")
+        while self.vehicle.location.global_relative_frame.alt > 0.1:
```

Comment:
Is this check reliable?

Reply:
During my testing in QGroundControl, I have not encountered an issue with the tolerance here. Obviously it can't be zero since the drone will never reach a zero altitude, so this has been a reliable check on the altitude during RTL mode.